Scalar scaling fix

roystgnr opened the pull request:
I believe this should fix #9450 if my understanding of the problem is correct.

The second commit here trips _GLIBCXX_DEBUG assertion failures on the perverse N_proc > N_elem + 1 case, verifying an attempt to access uninitialized data, and the third commit fixes that failure, so there's at least a little justification for my belief.

We should change that second commit to be more robust (NaNs in opt mode?) before merging this, but I'll wait for a bit of discussion on exactly how we should do so; for now I just want to make sure that Civet doesn't think I've broken anything else.

roystgnr committed:
const accessors for MooseVariableScalar::dof*

roystgnr committed:
MooseVariableScalar::reinit() non-semilocal case

If not all our scalar dof indices are semilocal:

1. Get as much semilocal data as we can.

2. Resize arrays if we're in DEBUG mode so that we'll get a nice
traceable scream-and-die if someone tries to access non-semilocal
data.

The behavior in 2 is temporary and should later be changed to reflect
whatever we decide on when discussing #9450, which this code was used
to investigate.

roystgnr committed:
Check have_scalar_contributions pre-contribute

In theory this should avoid the redundant/poorly-scaling assembly
communication discussed in #9450.

In practices this definitely avoids triggering usage of uninitialized
data in the perverse case discussed there.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/99279/) on e173806 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 0c2822da3529ba3d36a72b1b1b603b538da6643b`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9451/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9451/style.patch | git apply -v`

roystgnr committed:
Formatting fixes

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/99294/) on ef18030 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9451/site)

roystgnr commented:
1e-5 absolute error is kind of borderline; too small to be sure it's a bug, too big to ignore unless we can be sure it's not a solver tolerance issue.  I'll look into it.

roystgnr committed:
Avoid false positive failures in DEBUG mode

If we're trying to prevent users from inadvertently reading
uninitialized values, we don't want to also prevent users from
*setting* *initialized* values.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/99604/) on d5ad750 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 0c2822da3529ba3d36a72b1b1b603b538da6643b`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9451/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9451/style.patch | git apply -v`

roystgnr committed:
Formatting fixes

roystgnr committed:
Fix ScalarVariable postprocessor w/o variable

In cases where a SCALAR variable isn't ghosted on every processor,
this postprocessor needs to be able to query it from the processors
which do have it and then communicate.

roystgnr commented:
Do we have an "App tests debug" test that we can manually run?  That _GLIBCXX_DEBUG trick caught a corner-case bug in the ScalarVariable postprocessor, and it would be nice to ensure that there aren't any similar problems in user code.

roystgnr commented:
I can replicate the generalized_plane_strain_tm_contact.generalized_plane_strain_tm_contact bug, BTW, but I don't yet understand it.  I've ruled out "test is running with too low a solver tolerance", and I hope to soon confirm or deny "gold standard was generated with too low a solver tolerance", but if neither theory is correct I've got no leads.

permcody commented:
I'm not entirely sure that your change is necessarily a bug. We've tripped these tests several times over the last few years with diffs that look like yours. Sometimes it's just the differences in the way we calculate things that cause these issues. Many of these tests are very sensitive. Rather than you continuing to look for the reason perhaps we could tackle it from the other end (i.e. involve one of the test builders to look at the new solutions to see if they still look reasonable). I'm happy to tag someone...

roystgnr commented:
Yeah, I'd appreciate it.  I don't get any other regression test failures (after I rooted out a weird MKL red herring), but I'm not understanding this one.  The diffs are around 1e-5, in a field whose peak solution is around 2e3, so I'd normally laugh and chalk that up to solver tolerance, but futzing with tolerances hasn't fixed anything for me, and I'm loathe to just crank up a huge abs tolerance floor on a test case I'm unfamiliar with.

permcody commented:
@bwspenc, or @hchen139 maybe one of you can shed some light on this test. Looking at the history it looks like it has given us fits before. The L2 norm of the residual of some of the timesteps is 1e-5-1e-6 so I'm surprised it doesn't diff more often. I really don't want Roy to try to figure out why his change is only breaking one or two of these very sensitive test when the rest of the suite passes.

permcody commented:
@fdkong - previously skipped this test: https://github.com/idaholab/moose/commit/c88bc1a6ffeca97ae6785b5eb9463c98e5acbc47

permcody commented:
@bwspenc and @hchen139 - Can one of you take a look at the test results here and see if you think these diffs are significant. You can read the comments above and the original issue but we are stuck trying to figure out if we've broken something here. Unless you are completely sure about the results in this test, it might actually be a fix. Any help would be appreciated, we'd like to get his patch in.

bwspenc commented:
@hchen139 is looking into this.  I doubt this is a problem with the code.  It looks like the test is just a little too sensitive to small numerical issues.  We'll plan on just submitting a separate PR to improve the test, unless we hear otherwise.

roystgnr commented:
We should kick those tests to see if they work now that #9531 is in.

I've got to handle those robustness issues, too, so I'll do that shortly and we can just wait for that push to trigger new tests.

roystgnr committed:
MooseVariableScalar non-global support more robust

Changed the macro to match libstdc++.

Added NaN usage for when the libstdc++ checking isn't available.

roystgnr commented:
Ah, hell.  I just rebased on master rather than devel, didn't I...

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/102826/) on 480965b wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 669f0516bb387a2bc55900114d73ec47baad83e8`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9451/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9451/style.patch | git apply -v`

roystgnr commented:
That should fix it.  If this passes tests now then IMHO it's good to go.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/102835/) on 912f1ad wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 669f0516bb387a2bc55900114d73ec47baad83e8`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9451/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9451/style.patch | git apply -v`

roystgnr commented:
Oh, come on!  My formatting looks better before the demanded patch.

roystgnr committed:
Make whitespace uglier at clang-format insistance

andrsd commented on framework/src/postprocessors/ScalarVariable.C:
```diff
@@ -18,6 +18,9 @@
 #include "MooseVariableScalar.h"
 #include "SubProblem.h"
 
+// libMesh includes
```
cpt. Obvious.

andrsd commented on framework/src/postprocessors/ScalarVariable.C:
```diff
@@ -49,5 +52,14 @@ Real
+  if (dof >= dof_map.first_dof() && dof < dof_map.end_dof())
+    returnval = _var.sln()[_idx];
+
+  this->comm().min(returnval);
```
use `gatherMin(...)`

roystgnr replied:
Will do.

andrsd commented on framework/src/base/MooseVariableScalar.C:
```diff
@@ -23,6 +23,9 @@
 #include "libmesh/numeric_vector.h"
 #include "libmesh/dof_map.h"
 
+// C++
```
Remove this line, it is cpt. Obvious.

roystgnr replied:
```
roystgnr@leela:(roys_junk3) ~/git/moose-fresh
$ egrep -i '// (system|C\+\+)' framework/include/*/*.h | wc
     35     118    2112
roystgnr@leela:(roys_junk3) ~/git/moose-fresh
$ egrep -i '// libMesh includes' framework/include/*/*.h | wc
     70     210    4668
```

Should I get the other hundred instances while I'm at it?  That should probably be a separate PR.

I'd rather not remove any of them at all.  When I was removing unnecessary header dependencies, I discovered that dependency-tracking best practices (e.g. https://stackoverflow.com/a/2762596 ) were followed less often than not.  I'd like to have *some* comments that express "This isn't random order; someone put these headers below those headers for a reason, and here's what category they were sorted by".

andrsd replied:
I understand that grouping header files is a good thing, but I also think people already do that. That's why I think we do not need comments like "Put C++/libmesh headers here"...

roystgnr replied:
I still disagree, but I'd rather get this merged than argue, so I'll push one more commit stripping those two comments just as soon as these retests finish.

That said,
```
roystgnr@zappa:(scalar_scaling_fix) ~/git/moose/modules
$ git blame ../framework/src/base/MooseVariableScalar.C | tail -n +22 | head -n 5
e4501c55 (David Andrs       2012-02-06 08:48:49 +0000  22) // libMesh
ff8a0bdc (Cody Permann      2013-02-07 12:33:07 +0000  23) #include "libmesh/numeric_vector.h"
02f63040 (John W. Peterson  2015-11-03 10:18:43 -0700  24) #include "libmesh/dof_map.h"
e4501c55 (David Andrs       2012-02-06 08:48:49 +0000  25) 
b99fd8ac (Roy H. Stogner    2017-07-05 17:16:21 -0500  26) // C++
```

https://en.wikipedia.org/wiki/I_learned_it_by_watching_you!   ;-)

andrsd replied:
Well, originally the code looked like this:
```
// libMesh
#include "numeric_vector.h"
#include "dof_map.h"
```
Then, you guys in libmesh added "libmesh/", see `git show ff8a0bdc` for the change. At that point we should have dropped `// libMesh` comments, but they stayed. I really do think they are cpt. Obvious.  With the same logic, I do not think we need `// C++`

roystgnr committed:
Prefer MOOSE gatherMin(foo) to comm().min(foo)

Thanks to @andrsd for pointing out the API.

roystgnr commented:
Looks like we're still seeing exodiffs on the same test, even after #9531 cranked down the solver tolerance?  I'll see if I can still replicate.

bwspenc commented:
@roystgnr Maybe we need to also loosen up the exodiff tolerances a bit on that test.  I'll look at the results after this round of tests runs.

moosebuild commented:
Job [TBB Test](https://www.moosebuild.org/job/102866/) on 8ddc378 : invalidated by @bwspenc

retest after fixing problematic test

moosebuild commented:
Job [TBB Test](https://www.moosebuild.org/job/102866/) on 8ddc378 : invalidated by @bwspenc

retest after fixing problematic test

moosebuild commented:
Job [Pthreads Test](https://www.moosebuild.org/job/102865/) on 8ddc378 : invalidated by @bwspenc

retest after fixing problematic test

moosebuild commented:
Job [Modules debug](https://www.moosebuild.org/job/102863/) on 8ddc378 : invalidated by @bwspenc

retest after fixing problematic test

roystgnr committed:
Remove 'Cpt. Obvious' comments

Either permanently or temporarily, depending on whether @andrsd or I
are better with pistols at dawn.

roystgnr commented:
Tests passed, so this should be good to merge.  That last commit was just comment deletions.
